### ISSUE TITLE
'ru' translation has invalid string patterns

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/beers-law-lab#291]
Related to https://github.com/phetsims/beers-law-lab (add support for dynamic locale) ...

When cycling through the locales with Ctrl-I, the sim crashes when advancing to whatever locale is after "ro".

The error message and stack trace are not very helpful:

> assert.js:28 Uncaught Error: Assertion failed: maxWidth should be null (no constraint) or a positive number
    at window.assertions.assertFunction (assert.js:28:13)
    at Text.setMaxWidth (Node.ts:2822:15)
    at set maxWidth [as maxWidth] (Node.ts:2839:10)
    at NumberDisplay.ts:247:7
    at TinyProperty.emit (TinyEmitter.ts:95:9)
    at DerivedProperty._notifyListeners (ReadOnlyProperty.ts:304:23)
    at DerivedProperty.unguardedSet (ReadOnlyProperty.ts:255:14)
    at DerivedProperty.set (ReadOnlyProperty.ts:239:12)
    at DerivedProperty.getDerivedPropertyListener (DerivedProperty.ts:155:13)
    at TinyProperty.emit (TinyEmitter.ts:95:9)

**Comment by pixelzoom** [comment]
Looks like 'ru' is after 'ro'.  Looking through beers-law-lab_run.json, I see that 4 of the string patterns have been translated incorrectly, see below. Their value is the empty string , which will certainly cause problems with the old `StringUtils.format`.

Rossetta is supposed to prevent this. The most recent commit to this file is 8/9/17.

```
  "pattern.0formula.1name": {
    "value": "",

  "pattern.0label": {
    "value": "",

  "pattern.0percent": {
    "value": "",

  "pattern.0value.1units": {
    "value": "",
```

@jbphet please advise how to fix this string file.

**Comment by pixelzoom** [comment]
I temporarily edited the 'ru' string file, and added the correct English pattern strings. The assertion failure disappears. 

Then I verified that this specific string is causing the assertion failure:

```
  "pattern.0value.1units": {
    "value": "",
```

This is used in a couple of places as the `valuePattern` option to NumberDisplay, which seems to be having problems with the empty string.  It's like the usage in ConcentrationControl:

```
    options.numberDisplayOptions.valuePattern = new DerivedProperty(
      [ BeersLawLabStrings.pattern[ '0value' ][ '1unitsStringProperty' ], transform.unitsStringProperty ],
      ( pattern, unitsString ) => StringUtils.format( pattern, SunConstants.VALUE_NUMBERED_PLACEHOLDER, unitsString )
    );
```

Finally, if this is the translation that's currently published, I'd be surprised if the 'ru' version of the sim is usable.

**Comment by pixelzoom** [comment]
In https://github.com/phetsims/scenery-phet/issues/770, I fixed NumberDisplay so that it doesn't crash with an empty string for `valuePattern`.  

@jbphet the remaining work here is to fix the 'ru' translation. Please advise.

**Comment by jbphet** [comment]
@pixelzoom said above:

> ...4 of the string patterns have been translated incorrectly, see below. Their value is the empty string , which will certainly cause problems with the old StringUtils.format.
>
> Rossetta is supposed to prevent this.

I'm not sure Rosetta actually is designed to prevent this.  I'll can more investigation later if necessary, but I think Rosetta is designed to prevent translators from entering *incorrect* patterns, but removing a previously translated pattern seems like a reasonable use case to me.  The translator is essentially saying, "I've decided that I want to use the default value instead of what I'd previously entered".  When that happens, the strings should fall back to the value in the English file.  So, my initial evaluation here is that the problem lies in the string processing, which should fall back to values from the English strings file if there is a null string for the value in the translated file.

@pixelzoom - I'm handing this back to you to follow up on this thought, but you can send it back to me if you believe I'm off in the weeds.  Also, there *is* one previous value for the string `pattern.0formula.1name` that doesn't follow the pattern, and that somehow got through Rosetta, so something does seem a bit off with this.

**Comment by pixelzoom** [comment]
My understanding is that string processing only falls back to the English value when there is no translated value.

This string in the 'ru' file:

```
  "pattern.0value.1units": {
    "value": "",
```

... is most definitely not being not the absence of a value. It's the presence of a value, which happens to be the empty string. If Rosetta is responsible for writing the empty string when the translator decides that they want the English fallback, then that is likely the problem.

**Comment by jbphet** [comment]
I just performed a test of Rosetta (running locally on my machine, not the live version), and it definitely allows a user to delete a translation value and submit that change, which results in a zero-length string being submitted to the strings file in the babel repo.  This is apparently a use case that was never considered in the original design and implementation.

I've done a bunch of investigation on this, and there are some of the details below, but my conclusion thus far is this:  Rosetta has always behaved this way, but there was something in the build process that would cause a sim to fall back to English strings if a locale-specific value was a zero-length string, which is perhaps why this never caused obvious problems before.

I'll set up a more global issue to follow up on this.

Here are some more details of the investigation.  The balancing-act (BA) sim has a Gujarati (locale = gu) translation where the translator originally specified labels for the mystery masses, then set them all back to "".  In the published Gujaruati version of BA, which is version 1.1.27, the mystery masses look like this (they are using the English labels):

![image](https://user-images.githubusercontent.com/3082831/195933275-8ee1cbeb-f4bf-4725-ab12-d4124d256d29.png)

However, if I build the Gujarati version on master, they look like this:

![image](https://user-images.githubusercontent.com/3082831/195937771-22534a0c-1b9d-4dd4-b0b7-4719d0d2f5e4.png)

This seems sufficient to demonstrate that something has changed, but there are a some complications.  I checked out the 1.1 branch of BA and ran the unbuilt version, and it showed blank labels for the mystery masses, so the fallback seems specific to the built versions, which seems weird.  I'm not sure what to make of that, but I've spent a lot of time on this already and can't afford to figure out why the old built and unbuilt versions differ.

**Comment by jbphet** [comment]
I've created a more overreaching issue for this problem, see https://github.com/phetsims/chipper/issues/1343.  I'll leave this one open and assigned to me, and follow up once we've decided on how we want to address it.

**Comment by jbphet** [comment]
The string-handling code has now been changed so that it falls back to the English translation when it encounters a zero-length string, see https://github.com/phetsims/chipper/issues/1343 for details.

@pixelzoom - While the work in https://github.com/phetsims/chipper/issues/1343 isn't 100% complete, I believe the specific case that spawned this issue has now been addressed.  I tested the ru version of BLL and it seemed to run fine.  Can we close this issue?
